File: cane-csharp/cane-core/Runtime/Constraints/Physics.cs
using System;
using System.Collections.Generic;
using Cane.Animation;
using Cane.Format;

namespace Cane
{
    public readonly struct PhysicsEnvironment
    {
        public readonly float WindX, WindY, GravityX, GravityY;
        public PhysicsEnvironment(float windX, float windY, float gravityX, float gravityY)
        { WindX = windX; WindY = windY; GravityX = gravityX; GravityY = gravityY; }
        public static PhysicsEnvironment Default => new PhysicsEnvironment(1, 0, 0, 1);
    }
}
namespace Cane.Constraints
{
    internal sealed class PhysicsBudget
    {
        internal int Remaining = 1000000;
        internal void Use(string id)
        { if (Remaining == 0) throw Limit(id); Remaining--; }
        internal static RuntimeException Limit(string id) => new RuntimeException(RuntimeErrorCode.ResourceLimit, "apply", "Physics fixed-step budget is exhausted or cannot make progress.");
    }
    internal sealed class PhysicsState
    {
        internal bool Reset = true;
        internal float Ux, Uy, Cx, Cy, TipX, TipY, XOffset, YOffset, XLag, YLag, XVelocity, YVelocity;
        internal float RotateOffset, RotateLag, RotateVelocity, ScaleOffset, ScaleLag, ScaleVelocity, Remaining;
        internal PhysicsState Clone() => (PhysicsState)MemberwiseClone();
        internal bool IsFinite => Numeric.Finite(Ux) && Numeric.Finite(Uy) && Numeric.Finite(Cx) && Numeric.Finite(Cy) && Numeric.Finite(TipX) && Numeric.Finite(TipY)
            && Numeric.Finite(XOffset) && Numeric.Finite(YOffset) && Numeric.Finite(XLag) && Numeric.Finite(YLag) && Numeric.Finite(XVelocity) && Numeric.Finite(YVelocity)
            && Numeric.Finite(RotateOffset) && Numeric.Finite(RotateLag) && Numeric.Finite(RotateVelocity) && Numeric.Finite(ScaleOffset) && Numeric.Finite(ScaleLag) && Numeric.Finite(ScaleVelocity) && Numeric.Finite(Remaining);
    }
    internal sealed partial class PhysicsStore
    {
        internal readonly Dictionary<string, PhysicsState> States = new Dictionary<string, PhysicsState>(StringComparer.Ordinal);
        private string? animation;
        private float? time, rawTime;
        internal PhysicsStore Clone()
        { var result = new PhysicsStore { animation = animation, time = time, rawTime = rawTime }; foreach (var p in States) result.States.Add(p.Key, p.Value.Clone()); return result; }
        internal bool Reset(string? id = null)
        {
            if (id == null) { bool had = States.Count != 0; States.Clear(); animation = null; time = null; rawTime = null; return had; }
            if (!States.ContainsKey(id)) return false; States[id] = new PhysicsState(); return true;
        }
        internal PhysicsState State(string id)
        { if (!States.TryGetValue(id, out PhysicsState? result)) { result = new PhysicsState(); States.Add(id, result); } return result; }
        internal float Advance(Playback state, Pose pose, SamplingOptions sampling)
        {
            Entry? current = state.Tracks.TryGetValue(0, out Track? track) ? track.Current : null;
            float raw = current?.Time ?? state.DetachedTime, now = sampling.Time(raw); string? clip = current?.Clip?.Id;
            bool changed = animation != clip || rawTime.HasValue && raw < rawTime.Value;
            float? previous = changed ? null : time, previousRaw = changed ? null : rawTime;
            if (changed) States.Clear(); float delta = Math.Max(0, previous.HasValue ? now - previous.Value : now);
            animation = clip; time = now; rawTime = raw;
            if (current?.Clip != null && previousRaw.HasValue && raw > previousRaw.Value && current.Alpha * current.Progress >= 0.5f)
            {
                foreach (Json timeline in current.Clip.Raw["constraintTimelines"].Items)
                {
                    if (timeline.S("type") != "physics") continue; bool crossed = false;
                    foreach (Json key in timeline["keys"].Items)
                    {
                        if (!key.B("reset")) continue; float k = key.F("time") - current.Start;
                        if (current.Loop && current.Duration > 0)
                        { crossed |= Math.Floor(((double)previousRaw.Value - k) / current.Duration) < Math.Floor(((double)raw - k) / current.Duration); }
                        else crossed |= k > previousRaw.Value && k <= raw;
                    }
                    if (!crossed) continue; string id = timeline.S("constraintId");
                    if (id != "*") Reset(id);
                    else foreach (Json c in pose.Constraints) if (c.S("type") == "physics" && pose.IsActive(c.S("id"), "constraintIds") && pose.IsActive(c.S("boneId"), "boneIds")) Reset(c.S("id"));
                }
            }
            return delta;
        }
    }
    internal static class Physics
    {
        internal static void Solve(Pose pose, Json c, Affine root, PhysicsStore store, float delta, PhysicsEnvironment env, float rootScaleX, float rootScaleY, PhysicsBudget budget)
        {
            string id = c.S("id"), boneId = c.S("boneId"); if (!pose.IsActive(boneId, "boneIds")) return;
            PhysicsState s = store.State(id);
            float mix = c.F("mix", 1), fps = c.F("fps", 60), reference = pose.Data.ReferenceScale;
            if (mix == 0 || fps == 0 || !(reference > 0)) return;
            int bone = pose.Data.BoneIndex[boneId]; float length = Math.Max(pose.Data.BoneData[bone].Length, 0); Affine m = pose.World[bone];
            float step = 1 / fps, originX = m.Tx, originY = m.Ty, previousRemaining = s.Remaining;
            float remaining = s.Remaining + delta, planned = remaining; int count = 0;
            float x = c.F("x"), y = c.F("y"), rotate = c.F("rotate"), shear = c.F("shearX"), scaleX = c.F("scaleX");
            bool xe = x > 0, ye = y > 0, re = rotate > 0 || shear > 0, se = scaleX > 0;
            if (!s.Reset && (xe || ye || re || se))
            {
                if (!Numeric.Finite(step) || step <= 0 || !Numeric.Finite(remaining)) throw PhysicsBudget.Limit(id);
                while (planned >= step) { budget.Use(id); float next = planned - step; if (next == planned) throw PhysicsBudget.Limit(id); planned = next; count++; }
            }
            float interpolation = 0; s.Remaining = remaining; float? requiredLimit = 0; bool limitSaturated = false;
            if (s.Reset) { s.Reset = false; s.Ux = originX; s.Uy = originY; }
            else
            {
                float inertia = Numeric.Clamp(c.F("inertia", 1), 0, 1), limit = Math.Max(c.F("limit", 5000), 0);
                float xLimit = limit * delta * Math.Abs(rootScaleX), yLimit = limit * delta * Math.Abs(rootScaleY);
                float damping = (float)Math.Pow(Numeric.Clamp(c.F("damping", 1), 0, 1), 60 * step), acceleration = step / Math.Max(c.F("mass", 1), float.Epsilon), strength = Math.Max(c.F("strength", 100), 0);
                if (xe || ye)
                {
                    if (xe) { float movement = (s.Ux - originX) * inertia; ObserveLimit(movement, xLimit, delta * Math.Abs(rootScaleX), ref requiredLimit, ref limitSaturated); s.XOffset += Numeric.Clamp(movement, -xLimit, xLimit); s.Ux = originX; }
                    if (ye) { float movement = (s.Uy - originY) * inertia; ObserveLimit(movement, yLimit, delta * Math.Abs(rootScaleY), ref requiredLimit, ref limitSaturated); s.YOffset += Numeric.Clamp(movement, -yLimit, yLimit); s.Uy = originY; }
                    if (count > 0)
                    {
                        float wind = reference * c.F("wind"), gravity = reference * c.F("gravity");
                        float fx = (wind * env.WindX + gravity * env.GravityX) * rootScaleX, fy = (wind * env.WindY + gravity * env.GravityY) * rootScaleY;
                        float beforeX = s.XOffset, beforeY = s.YOffset;
                        for (int i = 0; i < count; i++)
                        {
                            if (xe) { s.XVelocity += (fx - s.XOffset * strength) * acceleration; s.XOffset += s.XVelocity * step; s.XVelocity *= damping; }
                            if (ye) { s.YVelocity -= (fy + s.YOffset * strength) * acceleration; s.YOffset += s.YVelocity * step; s.YVelocity *= damping; }
                        }
                        s.XLag = s.XOffset - beforeX; s.YLag = s.YOffset - beforeY; remaining = planned;
                    }
                    interpolation = Math.Max(1 - remaining / step, 0);
                    m = new Affine(m.A, m.B, m.C, m.D, m.Tx + (xe ? (s.XOffset - s.XLag * interpolation) * mix * x : 0), m.Ty + (ye ? (s.YOffset - s.YLag * interpolation) * mix * y : 0));
                }
                if (re || se)
                {
                    ObserveLimit(s.Cx - m.Tx, xLimit, delta * Math.Abs(rootScaleX), ref requiredLimit, ref limitSaturated);
                    ObserveLimit(s.Cy - m.Ty, yLimit, delta * Math.Abs(rootScaleY), ref requiredLimit, ref limitSaturated);
                    float axisAngle = Numeric.Atan2(m.B, m.A), tipX = Numeric.Clamp(s.Cx - m.Tx, -xLimit, xLimit), tipY = Numeric.Clamp(s.Cy - m.Ty, -yLimit, yLimit);
                    float rotateMix = (rotate + shear) * mix, co, si;
                    if (re)
                    {
                        float lag = s.RotateLag * Math.Max(1 - previousRemaining / step, 0);
                        float incoming = Numeric.Atan2(tipY + s.TipY, tipX + s.TipX) - axisAngle - (s.RotateOffset - lag) * rotateMix;
                        float wrapped = (incoming + Numeric.Pi) % (Numeric.Pi * 2); if (wrapped < 0) wrapped += Numeric.Pi * 2;
                        s.RotateOffset += (wrapped - Numeric.Pi) * inertia;
                        float angle = (s.RotateOffset - lag) * rotateMix + axisAngle; co = Numeric.Cos(angle); si = Numeric.Sin(angle);
                        if (se) { float worldLength = length * Solvers.Sx(m); if (worldLength > 0) s.ScaleOffset += (tipX * co + tipY * si) * inertia / worldLength; }
                    }
                    else
                    {
                        co = Numeric.Cos(axisAngle); si = Numeric.Sin(axisAngle);
                        float worldLength = length * Solvers.Sx(m) - s.ScaleLag * Math.Max(1 - previousRemaining / step, 0);
                        if (worldLength > 0) s.ScaleOffset += (tipX * co + tipY * si) * inertia / worldLength;
                    }
                    if (count > 0)
                    {
                        float fx = c.F("wind") * env.WindX + c.F("gravity") * env.GravityX, fy = c.F("wind") * env.WindY + c.F("gravity") * env.GravityY;
                        float lengthScale = length / reference, beforeRotate = s.RotateOffset, beforeScale = s.ScaleOffset;
                        for (int i = 0; i < count; i++)
                        {
                            if (se) { s.ScaleVelocity += (fx * co - fy * si - s.ScaleOffset * strength) * acceleration; s.ScaleOffset += s.ScaleVelocity * step; s.ScaleVelocity *= damping; }
                            if (re) { s.RotateVelocity -= ((fx * si + fy * co) * lengthScale + s.RotateOffset * strength) * acceleration; s.RotateOffset += s.RotateVelocity * step; s.RotateVelocity *= damping; }
                            if (re && i + 1 < count) { float angle = s.RotateOffset * rotateMix + axisAngle; co = Numeric.Cos(angle); si = Numeric.Sin(angle); }
                        }
                        s.RotateLag = s.RotateOffset - beforeRotate; s.ScaleLag = s.ScaleOffset - beforeScale; remaining = planned;
                    }
                    interpolation = Math.Max(1 - remaining / step, 0);
                }
                s.Remaining = remaining;
            }
            float rotationOffset = (s.RotateOffset - s.RotateLag * interpolation) * mix;
            if (rotate > 0) m = RotateRadians(m, rotationOffset * rotate, false);
            if (shear > 0) m = RotateRadians(m, rotationOffset * shear, true);
            if (scaleX > 0)
            {
                float scale = 1 + (s.ScaleOffset - s.ScaleLag * interpolation) * mix * scaleX, yScale = 1;
                if (c.S("scaleYMode", "none") == "uniform") yScale = scale;
                else if (c.S("scaleYMode") == "volume") { float absolute = Math.Abs(scale); yScale = absolute >= 0.7f ? 1 / absolute : 4 - 3.67347f * absolute; }
                m = new Affine(m.A * scale, m.B * scale, m.C * yScale, m.D * yScale, m.Tx, m.Ty);
            }
            s.Cx = originX; s.Cy = originY; s.TipX = length * m.A; s.TipY = length * m.B;
            if (!s.IsFinite || !m.IsFinite) throw new RuntimeException(RuntimeErrorCode.NonFinite, "apply", "Physics state became non-finite.", entityId: id);
            Solvers.WriteWorld(pose, bone, m, root);
            pose.Diagnostics[id] = new PhysicsConstraintDiagnostic(count, s, c.F("limit", 5000), requiredLimit, limitSaturated);
        }
        private static void ObserveLimit(float movement, float limit, float denominator, ref float? required, ref bool saturated)
        {
            float magnitude = Math.Abs(movement); if (magnitude > limit) saturated = true;
            if (magnitude != 0 && required.HasValue)
            { float candidate = magnitude / denominator; required = denominator == 0 || !Numeric.Finite(candidate) ? (float?)null : Math.Max(required.Value, candidate); }
        }
        private static Affine RotateRadians(Affine m, float angle, bool firstOnly)
        {
            float co = Numeric.Cos(angle), si = Numeric.Sin(angle);
            return new Affine(co * m.A - si * m.B, si * m.A + co * m.B, firstOnly ? m.C : co * m.C - si * m.D, firstOnly ? m.D : si * m.C + co * m.D, m.Tx, m.Ty);
        }
    }
}

File: cane-cpp/cane-core/src/constraints/physics.cpp
#include "physics.hpp"
#include "solvers.hpp"
#include <limits>

namespace cane::constraints {
namespace {
float sine(float angle) { return static_cast<float>(std::sin(static_cast<double>(angle))); }
float cosine(float angle) { return static_cast<float>(std::cos(static_cast<double>(angle))); }
float atan2(float y, float x) { return static_cast<float>(std::atan2(static_cast<double>(y), x)); }
float wrap_radians(float angle) {
    auto value = std::fmod(angle + pi, pi * 2); if (value < 0) value += pi * 2; return value - pi;
}
Affine rotate_radians(const Affine& m, float angle, bool first_only) {
    const auto co = cosine(angle), si = sine(angle);
    return {co * m.a - si * m.b, si * m.a + co * m.b, first_only ? m.c : co * m.c - si * m.d,
        first_only ? m.d : si * m.c + co * m.d, m.tx, m.ty};
}
void observe_limit(float movement, float limit, float denominator, PhysicsDiagnostic& diagnostic) {
    const auto magnitude = std::abs(movement);
    diagnostic.limit_saturated = diagnostic.limit_saturated || magnitude > limit;
    if (magnitude != 0 && diagnostic.required_limit) {
        const auto candidate = magnitude / denominator;
        if (denominator == 0 || !std::isfinite(candidate)) diagnostic.required_limit.reset();
        else diagnostic.required_limit = std::max(*diagnostic.required_limit, candidate);
    }
}
}
void solve_physics(Context& context, std::size_t index, PhysicsStore& store, const PhysicsStep& input, PhysicsBudget& budget) {
    using P = animation::ConstraintProperty;
    auto& pose = context.pose; pose.diagnostics.at(index) = std::monostate{};
    const auto& d = std::get<detail::PhysicsDefinition>(pose.model().constraints.at(index).definition);
    const auto p = pose.constraints[index]; const auto mix = p[P::mix], reference = pose.data().skeleton().reference_scale;
    if (!pose.active_constraints[index] || !pose.active_bones[d.bone] || mix == 0 || p.physics.fps == 0 || !std::isfinite(reference) || reference <= 0 || !input.environment.finite()) return;
    const auto& id = pose.data().catalog(RuntimeCatalogKind::constraint)[index].id;
    const auto failure = [&] { return Error(ErrorCode::non_finite, "apply", "Physics state or geometry became non-finite.", {}, id); };
    auto state = store.state(index).value_or(PhysicsState{}); auto available_budget = budget;
    const auto delta = std::isfinite(input.delta_seconds) ? std::max(input.delta_seconds, 0.0f) : 0;
    const auto length = std::max(pose.data().bones()[d.bone].length, 0.0f), step = 1 / static_cast<float>(p.physics.fps);
    auto matrix = pose.world[d.bone]; const auto origin_x = matrix.tx, origin_y = matrix.ty, previous_remaining = state.remaining;
    auto remaining = state.remaining + delta, planned_remaining = remaining;
    const auto xe = p.physics.x > 0, ye = p.physics.y > 0, re = p.physics.rotate > 0 || p.physics.shear_x > 0, se = p.physics.scale_x > 0;
    PhysicsDiagnostic diagnostic; diagnostic.configured_limit = p.physics.limit; diagnostic.required_limit = 0.0f;
    if (!state.reset && (xe || ye || re || se)) {
        if (!std::isfinite(step) || step <= 0 || !std::isfinite(remaining)) throw Error(ErrorCode::resource_limit, "apply", "Physics step or remaining time cannot be integrated.");
        while (planned_remaining >= step) {
            available_budget.consume(); const auto next = planned_remaining - step;
            if (next == planned_remaining) throw Error(ErrorCode::resource_limit, "apply", "Physics fixed-step subtraction cannot make progress.");
            planned_remaining = next; ++diagnostic.fixed_steps;
        }
    }
    float interpolation = 0; state.remaining = remaining;
    if (state.reset) { state.reset = false; state.ux = origin_x; state.uy = origin_y; }
    else {
        const auto inertia = std::clamp(p[P::inertia], 0.0f, 1.0f), limit = std::max(p.physics.limit, 0.0f);
        const auto root_x = input.root_scale_x, root_y = input.root_scale_y;
        const auto x_limit = limit * delta * std::abs(root_x), y_limit = limit * delta * std::abs(root_y);
        const auto damping = static_cast<float>(std::pow(static_cast<double>(std::clamp(p[P::damping], 0.0f, 1.0f)), 60 * step));
        const auto acceleration = step / std::max(p[P::mass], std::numeric_limits<float>::denorm_min()), strength = std::max(p[P::strength], 0.0f);
        if (xe || ye) {
            if (xe) {
                const auto movement = (state.ux - origin_x) * inertia;
                observe_limit(movement, x_limit, delta * std::abs(root_x), diagnostic);
                state.x_offset += std::clamp(movement, -x_limit, x_limit); state.ux = origin_x;
            }
            if (ye) {
                const auto movement = (state.uy - origin_y) * inertia;
                observe_limit(movement, y_limit, delta * std::abs(root_y), diagnostic);
                state.y_offset += std::clamp(movement, -y_limit, y_limit); state.uy = origin_y;
            }
            if (diagnostic.fixed_steps != 0) {
                const auto wind = reference * p[P::wind], gravity = reference * p[P::gravity]; const auto& env = input.environment;
                const auto force_x = (wind * env.wind_x + gravity * env.gravity_x) * root_x;
                const auto force_y = (wind * env.wind_y + gravity * env.gravity_y) * root_y;
                const auto previous_x = state.x_offset, previous_y = state.y_offset;
                for (std::uint32_t i = 0; i < diagnostic.fixed_steps; ++i) {
                    if (xe) { state.x_velocity += (force_x - state.x_offset * strength) * acceleration; state.x_offset += state.x_velocity * step; state.x_velocity *= damping; }
                    if (ye) { state.y_velocity -= (force_y + state.y_offset * strength) * acceleration; state.y_offset += state.y_velocity * step; state.y_velocity *= damping; }
                }
                state.x_lag = state.x_offset - previous_x; state.y_lag = state.y_offset - previous_y; remaining = planned_remaining;
            }
            interpolation = std::max(1 - remaining / step, 0.0f);
            if (xe) matrix.tx += (state.x_offset - state.x_lag * interpolation) * mix * p.physics.x;
            if (ye) matrix.ty += (state.y_offset - state.y_lag * interpolation) * mix * p.physics.y;
        }
        if (re || se) {
            const auto raw_x = state.cx - matrix.tx, raw_y = state.cy - matrix.ty;
            observe_limit(raw_x, x_limit, delta * std::abs(root_x), diagnostic); observe_limit(raw_y, y_limit, delta * std::abs(root_y), diagnostic);
            const auto axis_angle = atan2(matrix.b, matrix.a), tip_x = std::clamp(raw_x, -x_limit, x_limit), tip_y = std::clamp(raw_y, -y_limit, y_limit);
            const auto rotate_mix = (p.physics.rotate + p.physics.shear_x) * mix; float co, si;
            if (re) {
                const auto lag = state.rotate_lag * std::max(1 - previous_remaining / step, 0.0f);
                const auto incoming = atan2(tip_y + state.tip_y, tip_x + state.tip_x) - axis_angle - (state.rotate_offset - lag) * rotate_mix;
                state.rotate_offset += wrap_radians(incoming) * inertia;
                const auto force_angle = (state.rotate_offset - lag) * rotate_mix + axis_angle; co = cosine(force_angle); si = sine(force_angle);
                if (se) {
                    const auto world_length = length * x_scale(matrix);
                    if (world_length > 0) state.scale_offset += (tip_x * co + tip_y * si) * inertia / world_length;
                }
            } else {
                co = cosine(axis_angle); si = sine(axis_angle);
                const auto world_length = length * x_scale(matrix) - state.scale_lag * std::max(1 - previous_remaining / step, 0.0f);
                if (world_length > 0) state.scale_offset += (tip_x * co + tip_y * si) * inertia / world_length;
            }
            if (diagnostic.fixed_steps != 0) {
                const auto& env = input.environment;
                const auto force_x = p[P::wind] * env.wind_x + p[P::gravity] * env.gravity_x;
                const auto force_y = p[P::wind] * env.wind_y + p[P::gravity] * env.gravity_y;
                const auto length_scale = length / reference, previous_rotate = state.rotate_offset, previous_scale = state.scale_offset;
                for (std::uint32_t i = 0; i < diagnostic.fixed_steps; ++i) {
                    if (se) { state.scale_velocity += (force_x * co - force_y * si - state.scale_offset * strength) * acceleration; state.scale_offset += state.scale_velocity * step; state.scale_velocity *= damping; }
                    if (re) { state.rotate_velocity -= ((force_x * si + force_y * co) * length_scale + state.rotate_offset * strength) * acceleration; state.rotate_offset += state.rotate_velocity * step; state.rotate_velocity *= damping; }
                    if (re && i + 1 < diagnostic.fixed_steps) { const auto angle = state.rotate_offset * rotate_mix + axis_angle; co = cosine(angle); si = sine(angle); }
                }
                state.rotate_lag = state.rotate_offset - previous_rotate; state.scale_lag = state.scale_offset - previous_scale; remaining = planned_remaining;
            }
            interpolation = std::max(1 - remaining / step, 0.0f);
        }
        state.remaining = remaining;
    }
    const auto rotation_offset = (state.rotate_offset - state.rotate_lag * interpolation) * mix;
    if (p.physics.rotate > 0) matrix = rotate_radians(matrix, rotation_offset * p.physics.rotate, false);
    if (p.physics.shear_x > 0) matrix = rotate_radians(matrix, rotation_offset * p.physics.shear_x, true);
    if (p.physics.scale_x > 0) {
        const auto scale = 1 + (state.scale_offset - state.scale_lag * interpolation) * mix * p.physics.scale_x; float y_scale = 1;
        if (d.scale_y == detail::ScaleYMode::uniform) y_scale = scale;
        else if (d.scale_y == detail::ScaleYMode::volume) { const auto magnitude = std::abs(scale); y_scale = magnitude >= .7f ? 1 / magnitude : 4 - 3.67347f * magnitude; }
        matrix.a *= scale; matrix.b *= scale; matrix.c *= y_scale; matrix.d *= y_scale;
    }
    state.cx = origin_x; state.cy = origin_y; state.tip_x = length * matrix.a; state.tip_y = length * matrix.b;
    if (!state.finite() || !matrix.finite()) throw failure();
    diagnostic.translation_offset = cane::hypot(state.x_offset, state.y_offset); diagnostic.translation_speed = cane::hypot(state.x_velocity, state.y_velocity);
    diagnostic.rotation_offset_degrees = state.rotate_offset * radians_to_degrees; diagnostic.angular_speed_degrees = state.rotate_velocity * radians_to_degrees;
    diagnostic.scale_offset = state.scale_offset; diagnostic.scale_speed = state.scale_velocity;
    for (const auto value : {diagnostic.translation_offset, diagnostic.translation_speed, diagnostic.rotation_offset_degrees, diagnostic.angular_speed_degrees}) if (!std::isfinite(value)) throw failure();
    try { context.write_world(d.bone, matrix); } catch (const Error& error) { if (error.code == ErrorCode::non_finite) throw failure(); throw; }
    store.commit(index, state); budget = available_budget; pose.diagnostics[index] = diagnostic;
}
}
